File: 백준/Gold/30463. K－문자열/K－문자열.py
# https://www.acmicpc.net/problem/30463
import sys
input = sys.stdin.readline


def get_bit(x: str):
    bit = 0
    for a in map(int, x):
        bit |= 1 << a
    return bit


def find(x):
    cnt = 0
    for i in range(10):
        if (1 << i) & x:
            cnt += 1
    return cnt


n, k = map(int, input().split())
counts = [0] * 1024
for _ in range(n):
    counts[get_bit(input().rstrip())] += 1

total = 0
for i in range(1024):
    if find(i) == k:
        total += counts[i] * (counts[i]-1) >> 1
    for j in range(i+1, 1024):
        if find(i | j) == k:
            total += counts[i] * counts[j]
print(total)

File: 백준/Gold/30463. K－문자열/K－문자열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nk = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nk[0];
        int k = nk[1];

        int[] counts = new int[1024];
        for (int unused = 0; unused < n; unused++)
            counts[getBitmask(br.readLine())]++;

        long total = 0L;
        for (int i = 0; i < 1024; i++) {
            if (Integer.bitCount(i) == k)
                total += (long)counts[i] * (counts[i]-1) >> 1;

            for (int j = i+1; j < 1024; j++) {
                if (Integer.bitCount(i | j) == k)
                    total += (long)counts[i] * counts[j];
            }
        }

        System.out.println(total);
    }

    private static int getBitmask(String x) {
        int bit = 0;
        for (char c: x.toCharArray())
            bit |= 1 << (c - '0');
        return bit;
    }
}